### ISSUE TITLE
Better detection of originals

### ISSUE THREAD

**Issue by Sohalt** [sahib/rmlint#156]
I would like to have better detection which one of two identical files is the preferrable one to keep.
For example having a file `foo.png` and `foo (1).png` I would like to keep foo.png
The same goes for all the other ways of "incrementing a filename" (`foo_42.txt`, `bar{3}.md`, etc)
In general I would like to be able to specify a regex that decides which file should be kept.

**Comment by sahib** [comment]
In ``rmlint/develop`` you can do the following:

```bash
$ rmlint -S lad  #  This will sort paths first by length of the basename, then by abc, then by path depth.
```

So, the original will be the one with a shorter name, and if both have a number, the  one with the lower number and if still tied, the one which is higher in the filesystem hierachy. In your case that would be ``foo.png``. In 99% of all cases this should be sufficient. 

How would your regex version work? What's your exact proposal?

**Comment by Sohalt** [comment]
Using the length helps but is not sufficient for all use cases.
For the regex option I would like to be able to specify a regex that get's matched against the paths of a duplicate and the paths where the regex matches are kept and others deleted (or the other way around).

**Comment by sahib** [comment]
Note: We're looking into this, but there are some other todos that have to be popped from the queue before.

One problem I see is that a too common regex might work in one case but does the wrong thing otherwise. It probably takes some time to adjust the regex to be fit enough to do what you want.
Here's for example one that might work for your initial example by matching the "bad original":

```bash
'\w+\s*\(\d+\)\..*'  # "file.png" > "file (1).png" 
```
There are surely plenty more ways to write it.
Is there really a usecase that can't be done by being clever with ``-S``? Examples?

**Comment by Sohalt** [comment]
> Note: We're looking into this, but there are some other todos that have to be popped from the queue before.

Thanks!

> Is there really a usecase that can't be done by being clever with -S? Examples?

For example if I have a folder with images where the filenames are of the form `IMG_<some_number>.png` or `DSC<some_nuber>.jpg` and I have duplicates of those images where I have renamed the file to something more meaningful, i.e. `trees.png` and `lake.jpg` I would like to keep the renamed and in most cases more useful version and drop the generic IMG_something version. I think this is pretty much impossible to do by being clever with `-S`.

**Comment by sahib** [comment]
I've implemented a sketch of your idea:

https://github.com/sahib/rmlint/tree/feature/regex-sortcriteria

This ironically makes pretty much everything possible now by being clever with ``-S``,
but one might get the readability of perl in return. There are two new sortcriteria letters:

* ``r``: Make all *paths* original that match the regex.
* ``x``:  Make all *basenames* original that match the regex.

The uppercase letters ``R`` and ``X`` do the opposite thing.

Usage is like this:

```bash
$ rmlint -S 'r<DSC_\d+\.jpg>'   # Original should contain DSC.
$ rmlint -S 'x<\(DSC|IMG\).*>'  # Originals should not contain DSC or IMG.
```

You can also use more than one regular expression or sort by other criterias afterwards or before.
Testing is welcome.

**Comment by sahib** [comment]
Update: 

Some improvements regarding speed were made. 
It shouldn't make a big speed difference when using this feature now.

On another topic: Not sure if this should land in the ``2.4.0`` release. Actually we're in some sort of
lax featuren freeze now. Your opinion @SeeSpotRun?

**Comment by SeeSpotRun** [comment]
I'm happy enough to include it if you have time to update docs and nosetests, but it would probably make more sense to leave it in the develop branch until 2.5.0.
What about also (re-)adding a traversal regex?

**Comment by sahib** [comment]
Alright, we wanted to reduce the time between releases anyways. 
This feature will be merged *after* releasing ``2.4.0``.

Regarding traversal regexes: Still not a big fan. You couldn't easily do something like ``-S 'r<xyz>'`` before, but you can always use ``find [...] | rmlint -`` for more complex traversal filtering.

**Comment by sahib** [comment]
I merged this into ``develop``. Closing therefore.
